text/trace/0157-messaging-semantic-conventions.md

```diff
@@ -0,0 +1,184 @@
+| An OTEP proposing a set of attributes and conventions covering the scenarios in this document is approved and merged. | 10/29/2021 |
+| Proposed specification changes are verified by prototypes for the scenarios and examples below. | 11/15/2021 |
+| The [specification for messaging semantic conventions for tracing](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md) are updated according to the OTEP mentioned above and are declared [stable](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#stable). | 11/30/2021 |
+
+## Terminology
+
+### Message
+
+A "message" is a transport wrapper for the transfer of information between two
+or more parties. The information is a combination of data and metadata.
+Messages may be transferred directly between parties or via one or more
+intermediaries.  A message may carry annotations that are directed at
+intermediaries on the message path. Messages are uniquely identifiable.
+
+In the strict sense, a _message_ is an item of data that is sent to a specific
+destination, whereas an _event_ is a signal emitted by a component upon
+reaching a given state. This document is agnostic of those differences and uses
+the term "message" in a wider sense to cover both concepts.
+
+### Producer
+
+The "producer" is a specific instance, process or device that creates and
+publishes a message. "Publishing" is the process of sending a message or batch
+to the intermediary or consumer.
+
+### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new events.
+
+The consumer receives, processes, and settles a message. "Receiving" is the
+process of obtaining a message from the intermediary, "processing" is the
+process of acting on the information a message contains, "settling" is the
+process of notifying an intermediary that a message was processed successfully.
+
+### Intermediary
+
+An "intermediary" receives a message to forward it to the next receiver, which
+might be another intermediary or a consumer.
+
+## Scenarios
+
+Producing and consuming a message involves five stages:
+
+```
+PRODUCER
+
+Create
+  |                            CONSUMER
+  v        +--------------+                   
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+       |
+                  ^               v
+                  .            Process
+                  .               |
+                  .               v
+                  . . . . . .  Settle
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed successfully.
```

Comment:
The settlement phase is optional.
For example, in rabbitMQ, consumer client can be configured with the `noAck` flag which means the settlement is automatically applied by the broker:

noAck (boolean): if true, the broker won’t expect an acknowledgement of messages delivered to this consumer; i.e., it will dequeue messages as soon as they’ve been sent down the wire. Defaults to false (i.e., you will be expected to acknowledge messages).

Pub-Sub systems (for example: mqtt with QOS 0/1 and amazon SNS) also do not, or can be configured to not apply any application-level settlement as far as I know. once the message is received by the consumer, it is considered settled.

Reply:
On this topic, Quarkus has the ability to acknowledge a message before it's actually processed. Not sure how we can represent it in the flow above, but we need to think about these alternative flows

Reply:
Thanks. I added those possible variations (settling before processing, and settling bundled with receiving).